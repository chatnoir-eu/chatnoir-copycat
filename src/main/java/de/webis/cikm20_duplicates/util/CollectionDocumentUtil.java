package de.webis.cikm20_duplicates.util;

import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.time.temporal.ChronoUnit;

import org.apache.commons.io.IOUtils;

import de.webis.WebisUUID;
import de.webis.cikm20_duplicates.spark.SparkEnrichRelevanceTransferPairs;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.index.transform.JsoupStringTransform;
import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;

public class CollectionDocumentUtil {

	@SneakyThrows
	public static void main(String[] args) {
//		// S3: 1.0 (Should be valid)
//		String firstId = "clueweb12-0001wb-23-26353", secondId = "clueweb12-0005wb-24-06245";
		
//		// S3: 0.0 (Should be invalid)
//		String firstId = "clueweb12-0912wb-53-11472", secondId = "clueweb12-1000tw-11-07961";
		
		// S3: 0.0 (Should be invalid)
//		String firstId = "clueweb12-0303wb-10-02360", secondId = "clueweb12-0715wb-02-10722";

//		// S3: 0.0 (Should be invalid)
//		String firstId = "clueweb12-0303wb-35-15203", secondId = "clueweb12-0816wb-90-18920";
		
		// S3: 1.0 (Should be valid)
//		String firstId = "clueweb12-0406wb-47-22712", secondId = "clueweb12-1305wb-37-22239";
		
//		// S3: 1.0 (Should be valid)
//		String firstId = "clueweb12-0308wb-07-19989", secondId = "clueweb12-1505wb-58-23691";
		
//		// S3: 0.899 (Should be valid)
//		String firstId = "clueweb12-0810wb-02-19724", secondId = "clueweb12-0811wb-20-10663";
		
		
//		// S3: 0.99 (Should be valid)
//		String firstId = "clueweb12-0807wb-51-02126", secondId = "clueweb12-0810wb-43-15854";
		
//		// S3: 0.79 (should be invalid)
//		String firstId = "clueweb12-1313wb-20-00368", secondId = "clueweb12-1313wb-24-32697";
		
//		// S3: 0.94 (should be valid)
//		String firstId = "clueweb12-0805wb-13-09655", secondId = "clueweb12-0809wb-63-03994";
		
//		// S3: 0.97 (should be valid)
//		String firstId = "clueweb12-0810wb-79-00932", secondId = "clueweb12-0811wb-54-18364";
		
//		// S3: 0.93 (should be valid)
//		String firstId = "clueweb12-1600tw-18-12106", secondId = "clueweb12-1900tw-30-12707";
		
//		// S3: 0.79 (should be invalid)
//		String firstId = "clueweb12-1018wb-50-14351", secondId = "clueweb12-1019wb-33-29258";
		
//		System.out.println("Retrieve " + firstId + " located at: " + chatNoirURL(firstId) +"&plain");
//		CollectionDocument a = loadCollectionDocument(firstId);

//		System.out.println("Retrieve " + secondId + " located at: " + chatNoirURL(secondId) +"&plain");
//		CollectionDocument b = loadCollectionDocument(secondId);
		
//		System.out.println("Calculate S3-Score between " + firstId + " and " + secondId);
//		System.out.println(SparkEnrichRelevanceTransferPairs.s3Score(a,b));
		
		// {"equivalentDocuments": ["<urn:uuid:59a4637d-976f-4a5c-9861-cd5f3fc6c737>","<urn:uuid:be66ee99-8ee6-46ce-9e96-de6067eff76a>","<urn:uuid:ecffb099-540a-4176-8a97-7152a0576383>"],"hash":[-244842496, 47039, 11927603, 14439680]}
		//String firstId = "<urn:uuid:59a4637d-976f-4a5c-9861-cd5f3fc6c737>";
		//String secondId = "<urn:uuid:be66ee99-8ee6-46ce-9e96-de6067eff76a>";
		
		//{"equivalentDocuments": ["<urn:uuid:84606826-df14-4f0d-94cc-7d828f880949>","<urn:uuid:a1a75ce9-255f-4acd-83f9-30afd4fa6631>"],"hash":[-889782272, 50871, 14287080, 7504640]}
		// String firstId = "<urn:uuid:84606826-df14-4f0d-94cc-7d828f880949>";
		// String secondId = "<urn:uuid:a1a75ce9-255f-4acd-83f9-30afd4fa6631>";
		
		//{"equivalentDocuments": ["<urn:uuid:aa833e3e-0033-4937-b499-14e26d972404>","<urn:uuid:fb1ec340-b003-4dad-aa02-9a3ebe34b75f>"],"hash":[1527775232, 344, 13172802, 9114880]}
		// String firstId = "<urn:uuid:aa833e3e-0033-4937-b499-14e26d972404>";
		// String secondId = "<urn:uuid:fb1ec340-b003-4dad-aa02-9a3ebe34b75f>";
		
		//{"equivalentDocuments": ["<urn:uuid:177caa19-fd12-48dd-b85f-aa9151e1048e>","<urn:uuid:c74bf9bb-c949-47a1-bd4c-967e528a3e0d>","<urn:uuid:fd3341db-5dbb-45fe-b5c3-7aa1f91527a6>"],"hash":[367919104, 13594, 1900705, 413184]}
		String firstId = "<urn:uuid:177caa19-fd12-48dd-b85f-aa9151e1048e>";
		String secondId = "<urn:uuid:fd3341db-5dbb-45fe-b5c3-7aa1f91527a6>";
		
		System.out.println(chatNoirURL("commoncrawl", firstId, "cc1704"));
		CollectionDocument a = loadCollectionDocument(firstId, new URL(chatNoirURL("commoncrawl", firstId, "cc1704")));
		System.out.println(chatNoirURL("commoncrawl", secondId, "cc1704"));
		CollectionDocument b = loadCollectionDocument(secondId, new URL(chatNoirURL("commoncrawl", secondId, "cc1704")));
		
		System.out.println(SparkEnrichRelevanceTransferPairs.s3Score(a,b));
	}
	
	public static final RetryPolicy<String> RETRY_FINAL = new RetryPolicy<String>()
			.handle(Exception.class)
			.withBackoff(3, 60, ChronoUnit.SECONDS)
			.withMaxRetries(3);	
	
	@SneakyThrows
	public static CollectionDocument loadCollectionDocument(String documentId) {
		return loadCollectionDocument(new URL(chatNoirURL(documentId)), documentId, RETRY_FINAL);
	}
	

	public static CollectionDocument loadCollectionDocument(String documentId, URL chatNoirUrl) {
		return loadCollectionDocument(chatNoirUrl, documentId, RETRY_FINAL);
	}
	
	public static CollectionDocument loadCollectionDocument(URL url, String documentId, RetryPolicy<String> retryPolicy) {
		String documentContent = loadRawDocumentFromChatnoir(url, retryPolicy);
		
		JsoupStringTransform stringTransform = new JsoupStringTransform();
		return CollectionDocument.collectionDocument(stringTransform.apply(documentContent), documentId);
	}
	
	public static String loadRawDocumentFromChatnoir(URL url, RetryPolicy<String> retryPolicy) {
		return Failsafe.with(retryPolicy).get(() -> IOUtils.toString(url, StandardCharsets.UTF_8));
	}

	private static String webisUUID(String documentId) {
		return webisUUID(longChatNoirId(documentId), documentId);
	}
	
	private static String webisUUID(String prefix, String documentId) {
		return new WebisUUID(prefix)
				.generateUUID(documentId).toString();
	}
	
	private static String longChatNoirId(String documentId) {
		if(documentId.startsWith("clueweb09")) {
			return "clueweb09";
		} else if (documentId.startsWith("clueweb12")) {
			return "clueweb12";
		}
		
		throw new RuntimeException("ID '" + documentId + "' is not supported.");
	}
	
	private static String shortChatNoirId(String documentId) {
		if (documentId.startsWith("clueweb09")) {
			return "cw09";
		} else if (documentId.startsWith("clueweb12")) {
			return "cw12";
		}
		
		throw new RuntimeException("ID '" + documentId + "' is not supported.");
	}
	
	public static String chatNoirURL(String documentId) {
		return "https://chatnoir.eu/cache?uuid=" + webisUUID(documentId) + "&index="+ shortChatNoirId(documentId) +"&raw";
	}
	
	public static String chatNoirURL(String prefix, String documentId, String index) {
		return "https://chatnoir.eu/cache?uuid=" + webisUUID(prefix, documentId) + "&index="+ index +"&raw";
	}
}
