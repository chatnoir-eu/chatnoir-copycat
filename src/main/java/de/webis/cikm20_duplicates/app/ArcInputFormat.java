package de.webis.cikm20_duplicates.app;

import java.io.IOException;
import java.io.InputStream;
import java.util.Collections;
import java.util.Iterator;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FSDataInputStream;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.mapreduce.InputSplit;
import org.apache.hadoop.mapreduce.JobContext;
import org.apache.hadoop.mapreduce.RecordReader;
import org.apache.hadoop.mapreduce.TaskAttemptContext;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.input.FileSplit;
import org.archive.io.ArchiveRecord;
import org.archive.io.arc.ARCReader;
import org.archive.io.arc.ARCReaderFactory;
import org.archive.io.arc.ARCRecord;

import com.amazonaws.services.s3.model.AmazonS3Exception;

import lombok.SneakyThrows;

public class ArcInputFormat extends FileInputFormat<LongWritable, ARCRecord> {
	@Override
	public RecordReader<LongWritable, ARCRecord> createRecordReader(InputSplit split, TaskAttemptContext context)
			throws IOException, InterruptedException {
		return new ArcRecordReader();
	}

	@Override
	protected boolean isSplitable(JobContext context, Path filename) {
		return false;
	}

	public static class ArcRecordReader extends RecordReader<LongWritable, ARCRecord> {

		private LongWritable key = null;
		private ARCRecord value = null;
		private InputStream in;
		private Iterator<ArchiveRecord> iter;

		@Override
		public void close() throws IOException {
			if (in != null) {
				in.close();
			}
		}

		@Override
		public LongWritable getCurrentKey() throws IOException, InterruptedException {
			return key;
		}

		@Override
		public ARCRecord getCurrentValue() throws IOException, InterruptedException {
			return value;
		}

		@Override
		public float getProgress() throws IOException, InterruptedException {
			return 0.0f;
		}

		@Override
		public void initialize(InputSplit genericSplit, TaskAttemptContext context)
				throws IOException, InterruptedException {
			FileSplit split = (FileSplit) genericSplit;
			Configuration job = context.getConfiguration();
			final Path file = split.getPath();

			// open the file and seek to the start of the split
			FileSystem fs = file.getFileSystem(job);
			FSDataInputStream fileIn = fs.open(split.getPath());

			fileIn.seek(split.getStart());
			in = fileIn;

			initialize(file.toString(), fileIn);
		}

		public void initialize(String path, InputStream is) {
			key = null;
			value = null;
			iter = MyARCReaderFactory.getIteratorOrEmptyIterator(new Path(path), is);
		}

		@Override
		public boolean nextKeyValue() throws IOException, InterruptedException {
			try {
				if (!iter.hasNext()) {
					key = null;
					value = null;
	
					return false;
				} else {
					value = (ARCRecord) iter.next();
	
					if (value != null) {
						if (key == null) {
							key = new LongWritable();
						}
						key.set(value.getPosition());
					}
	
					return true;
	
				}
			} catch (Exception e) {
//				FIXME: REMOVE THIS
				return false;
			}
		}
	}

	public static class MyARCReaderFactory extends ARCReaderFactory {
		private static MyARCReaderFactory factory = new MyARCReaderFactory();

		@SneakyThrows
		public static Iterator<ArchiveRecord> getIteratorOrEmptyIterator(Path path, InputStream is) {
			try {
				return ((ARCReader) factory.getArchiveReader(path.toString(), is, Boolean.TRUE)).iterator();
			} catch (Exception e) {
//				FIXME: REMOVE THIS
				return Collections.emptyIterator();
			}
		}
	}
}
