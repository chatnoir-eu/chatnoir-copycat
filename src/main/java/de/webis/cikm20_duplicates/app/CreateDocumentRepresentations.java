package de.webis.cikm20_duplicates.app;

import java.io.Serializable;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.compress.BZip2Codec;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import de.webis.chatnoir2.indexer.util.ContentExtractor;
import de.webis.chatnoir2.indexer.util.LangDetector;
import de.webis.chatnoir2.mapfile_generator.warc.WarcRecord;
import de.webis.cikm20_duplicates.spark.SparkCanonicalLinkGraphExtraction;
import de.webis.cikm20_duplicates.spark.SparkCreateSourceDocuments;
import de.webis.cikm20_duplicates.util.SourceDocuments.DocumentWithFingerprint;
import de.webis.cikm20_duplicates.util.warc.WARCParsingUtil;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import lombok.SneakyThrows;
import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.ArgumentParserException;
import net.sourceforge.argparse4j.inf.Namespace;

public class CreateDocumentRepresentations {

	public static void main(String[] args) {
		Namespace parsedArgs = validArgumentsOrNull(args);

		if (parsedArgs == null) {
			return;
		}

		try (JavaSparkContext context = context()) {
			DocumentToTextTransformation transformation = transformation(parsedArgs);
			JavaPairRDD<LongWritable, WarcRecord> records = WARCParsingUtil.records(context, parsedArgs);
			JavaRDD<CollectionDocument> parsedDocuments = records.map(i -> transformToCollectionDocument(i._2(), transformation)).filter(i -> i != null);
			
			if (parsedDocuments.getNumPartitions() < 100) {
				parsedDocuments = parsedDocuments.repartition(parsedDocuments.getNumPartitions()*100);
			}

			JavaRDD<DocumentWithFingerprint> fingerprints = SparkCreateSourceDocuments.fingerprintAllDocuments(null,
					parsedDocuments, SparkCreateSourceDocuments.PRODUCTION_FINGERPRINTS);

			fingerprints.filter(i -> i != null).map(i -> i.toString())
					.saveAsTextFile(parsedArgs.getString(ArgumentParsingUtil.ARG_OUTPUT), BZip2Codec.class);
		}
	}
	
	static DocumentToTextTransformation transformation(Namespace parsedArgs) {
		//FIXME
		throw new RuntimeException("FIXME: implement and test this...");
	}

	@SneakyThrows
	public static CollectionDocument transformToCollectionDocument(WarcRecord record, DocumentToTextTransformation transformation) {
		if (record == null) {
			return null;
		}

		Map<String, String> header = lowercasedHeaders(record);
		String contentBody = record.getContent();

		if (contentBody.getBytes().length > 1024 * 1024 || !isWarcResponse(record)) {
			// ignore large files and non-responses
			return null;
		}

		if(transformation == null) {
			transformation = DocumentToTextTransformation.DEFAULT;
		}
		
		try {
			return transformToCollectionDocument(header, Jsoup.parse(contentBody), transformation);
		} catch (Exception e) {
			return null;
		}
	}
	
	public static boolean isWarcResponse(WarcRecord record) {
		return record != null && record.getRecordType() != null && "response".equalsIgnoreCase(record.getRecordType().trim());
	}

	@SuppressWarnings("serial")
	public static interface DocumentToTextTransformation extends Serializable {
		public String transform(Document doc);
		public static final LangDetector LANG_DETECTOR = langDetector();
		
		@SneakyThrows
		static LangDetector langDetector() {
			return new LangDetector();
		}
		
		public static final DocumentToTextTransformation DEFAULT = new DocumentToTextTransformation() {
			@Override
			public String transform(Document doc) {
				return doc.text();
			}
		};
		
		public static final DocumentToTextTransformation MAIN_CONTENT_EXTRACTION = new DocumentToTextTransformation() {
			
			@Override
			@SneakyThrows
			public String transform(Document doc) {
				String html = doc.toString();
				String lang = LANG_DETECTOR.detect(html);

	            if (lang != null && lang.equalsIgnoreCase("en")) {
	                return ContentExtractor.extract(html, "en");
	            } else {
	                return ContentExtractor.extract(html, lang, "en");
	            }

			}
		};
	}

	@SneakyThrows
	private static CollectionDocument transformToCollectionDocument(Map<String, String> header, Document doc, DocumentToTextTransformation transformation) {
		String id = header.get("warc-trec-id");
		if (id == null || id.isEmpty()) {
			id = header.get("warc-record-id");
		}

		String targetUri = header.get("warc-target-uri");

		CollectionDocument ret = CollectionDocument.collectionDocument(transformation.transform(doc), id);
		try {
			ret.setUrl(new URL(targetUri));
		} catch (Exception e) {}
		ret.setCanonicalUrl(SparkCanonicalLinkGraphExtraction.extractCanonicalLinkOrNull(targetUri, doc));
		ret.setCrawlingTimestamp(header.get("warc-date"));

		return ret;
	}

	public static Map<String, String> lowercasedHeaders(WarcRecord record) {
		return lowercasedHeaders(record.getHeader().getHeaderMetadata());
	}
	
	public static Map<String, String> lowercasedHeaders(Map<String, String> headers) {
		return new HashMap<>(headers.entrySet().stream()
				.filter(i -> i != null && i.getKey() != null && i.getValue() != null)
				.collect(Collectors.toMap(i -> i.getKey().trim().toLowerCase(), i -> i.getValue())));
	}

	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName(ArgumentParsingUtil.TOOL_NAME + ": CreateDocumentRepresentations");

		return new JavaSparkContext(conf);
	}

	static Namespace validArgumentsOrNull(String[] args) {
		ArgumentParser parser = argParser();

		try {
			return parser.parseArgs(args);
		} catch (ArgumentParserException e) {
			parser.handleError(e);
			return null;
		}
	}

	private static ArgumentParser argParser() {
		ArgumentParser ret = ArgumentParsers.newFor(ArgumentParsingUtil.TOOL_NAME + ": CreateDocumentRepresentations")
				.addHelp(Boolean.TRUE).build();

		ret.addArgument("-i", "--" + ArgumentParsingUtil.ARG_INPUT).required(Boolean.TRUE).help(
				"The input path that is passed to JavaSparkContext.hadoopFile to extract Documents from warc files. E.g. 's3a://corpus-clueweb09'.");

		ret.addArgument("-o", "--" + ArgumentParsingUtil.ARG_OUTPUT).required(Boolean.TRUE)
				.help("The resulting document representations are stored under this location.");

		ret.addArgument("-f", "--" + ArgumentParsingUtil.ARG_FORMAT).required(Boolean.TRUE)
				.choices(ArgumentParsingUtil.ALL_INPUT_FORMATS);

		return ret;
	}
}
