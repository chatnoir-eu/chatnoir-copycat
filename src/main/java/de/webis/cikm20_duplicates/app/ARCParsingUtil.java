package de.webis.cikm20_duplicates.app;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import org.apache.commons.httpclient.Header;
import org.apache.hadoop.io.LongWritable;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.archive.io.arc.ARCReader;
import org.archive.io.arc.ARCRecord;

import de.webis.cikm20_duplicates.app.ArcInputFormat.MyARCReaderFactory;
import net.sourceforge.argparse4j.inf.Namespace;

public class ARCParsingUtil {

	public static JavaPairRDD<LongWritable, ARCRecord> records(JavaSparkContext jsc, Namespace parsedArgs) {
		return records(jsc, parsedArgs.getString(ArgumentParsingUtil.ARG_INPUT));
	}
	
	public static JavaPairRDD<LongWritable, ARCRecord> records(JavaSparkContext jsc, String input) {
		return jsc.newAPIHadoopFile(input, ArcInputFormat.class, LongWritable.class, ARCRecord.class, jsc.hadoopConfiguration());
	}

	public static String extractURL(ARCRecord record) {
		if(record == null) {
			return "NOT-AVAILABLE";
		}
		
		 Header[] headers = record.getHttpHeaders();
		 if(headers == null) {
			 return "NOT-AVAILABLE"; 
		 }
		 
		 for(Header header: headers) {
			 if("x_commoncrawl_OriginalURL".equalsIgnoreCase(header.getName())) {
				 return header.getValue();
			 }
		 }
		 
		 return "ORIGINAL_URL_NOT_AVAILABLE";
	}

	public static List<String> extractAllURLs(String file, InputStream is) {
		ARCReader reader = MyARCReaderFactory.get(new org.apache.hadoop.fs.Path(file), is);
		List<String> ret = new ArrayList<>();
		
		reader.forEach(i -> ret.add(extractURL((ARCRecord) i)));
		
		return ret;
	}
}
