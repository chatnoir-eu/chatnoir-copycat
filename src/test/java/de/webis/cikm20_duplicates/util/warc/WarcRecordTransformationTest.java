package de.webis.cikm20_duplicates.util.warc;

import java.net.URL;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Test;
import org.mockito.Mockito;

import de.webis.chatnoir2.mapfile_generator.warc.WarcHeader;
import de.webis.chatnoir2.mapfile_generator.warc.WarcRecord;
import de.webis.cikm20_duplicates.app.CreateDocumentRepresentations;
import de.webis.cikm20_duplicates.app.CreateDocumentRepresentations.DocumentToTextTransformation;
import de.webis.cikm20_duplicates.spark.SparkIntegrationTestBase;
import de.webis.cikm20_duplicates.util.CollectionDocumentUtil;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import de.webis.cikm20_duplicates.util.CollectionDocumentUtil.EsDocumentResolver;
import de.webis.cikm20_duplicates.util.CollectionDocumentUtil.HdfsMapFileDocumentResolver;
import lombok.SneakyThrows;

public class WarcRecordTransformationTest extends SparkIntegrationTestBase {
	@Test
	public void testWithNullInput() {
		Assert.assertNull(CreateDocumentRepresentations.transformToCollectionDocument(null, null));
	}
	
	@Test
	public void approveTransformationOfClueWebRecord() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-TREC-ID", "my-id-1");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		WarcRecord record = record(headers, "my-main-content", "RESPONSE");
		
		CollectionDocument actual = transformToCollectionDocument(record);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void approveTransformationOfCommonCrawlRecord() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Record-ID", "my-id-2");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		WarcRecord record = record(headers, "<!DOCTYPE html>\n" + 
				"<html lang=\"de\">\n" + 
				"  <head>\n" + 
				"    <meta charset=\"utf-8\">\n" + 
				"    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n" + 
				"    <title>Titel</title>\n" + 
				"    <link rel=\"canonical\" href=\"https://example.com/test-123/\" />" +
				"  </head>\n" + 
				"  <body>Test 1 2 3\n" + 
				"\n" + 
				"  </body>\n" + 
				"</html>",
				"REsponse");
		
		CollectionDocument actual = transformToCollectionDocument(record);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void approveTransformationOfCommonCrawlRecordWithMainContentExtraction() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Record-ID", "my-id-2");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		WarcRecord record = record(headers, "<!DOCTYPE html>\n" + 
				"<html lang=\"de\">\n" + 
				"  <head>\n" + 
				"    <meta charset=\"utf-8\">\n" + 
				"    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n" + 
				"    <title>Titel</title>\n" + 
				"    <link rel=\"canonical\" href=\"https://example.com/test-123/\" />" +
				"  </head>\n" + 
				"  <body>Test 1 2 3\n" + 
				"\n" + 
				"  </body>\n" + 
				"</html>",
				"REsponse");
		
		CollectionDocument actual = transformToCollectionDocument(record, DocumentToTextTransformation.MAIN_CONTENT_EXTRACTION);
		
		Approvals.verifyAsJson(actual);
	}

	@Test
	@SneakyThrows
	public void approveTransformationOfExistingCommonCrawlRecord() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Record-ID", "my-id-2");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		String documentContent = CollectionDocumentUtil.loadRawDocumentFromChatnoir(new URL("https://chatnoir.eu/cache?uuid=92105ce9-2938-5b89-a191-7a82fe5d8816&index=cc1704&raw"), CollectionDocumentUtil.RETRY_FINAL);
		WarcRecord record = record(headers, documentContent, "REsponse");
		
		CollectionDocument actual = transformToCollectionDocument(record, DocumentToTextTransformation.DEFAULT);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	@SneakyThrows
	public void approveTransformationOfExistingCommonCrawlRecord2REMOVE() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Record-ID", "my-id-2");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		String documentContent = CollectionDocumentUtil.loadRawDocumentFromChatnoir(new URL("https://chatnoir.eu/cache?uuid=6336be0b-8971-50a2-9c8d-7a1f19cd1b66&index=cw09&raw"), CollectionDocumentUtil.RETRY_FINAL);
		WarcRecord record = record(headers, documentContent, "REsponse");
		
		CollectionDocument actual = transformToCollectionDocument(record, DocumentToTextTransformation.MAIN_CONTENT_EXTRACTION);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	@SneakyThrows
	public void approveTransformationOfExistinClueWebDocumentWithMainContentFromChatNoir() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Record-ID", "my-id-2");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		String documentContent = new EsDocumentResolver().loadCollectionDocument("clueweb09-en0000-00-00009").getContent();
		WarcRecord record = record(headers, documentContent, "REsponse");
		
		CollectionDocument actual = transformToCollectionDocument(record, DocumentToTextTransformation.MAIN_CONTENT_EXTRACTION);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	@SneakyThrows
	public void approveTransformationOfNonExistinClueWebDocumentWithMainContentFromChatNoir() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Record-ID", "my-id-2");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		CollectionDocument doc = new EsDocumentResolver().loadCollectionDocument("clueweb09-NON-EXISTING-en0000-00-00009");
		
		Assert.assertNull(doc);
	}

	@Test
	@SneakyThrows
	public void approveTransformationOfExistinClueWebDocumentFromHDFS() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Record-ID", "my-id-2");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		CollectionDocument doc = new HdfsMapFileDocumentResolver("cw09", "clueweb09").loadCollectionDocument("clueweb09-en0000-00-00009");
		WarcRecord record = record(headers, doc.getContent(), "REsponse");
		
		CollectionDocument actual = transformToCollectionDocument(record, DocumentToTextTransformation.DEFAULT);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	@SneakyThrows
	public void approveTransformationOfExistinClueWebDocumentFromHDFSWithCanonicalUrl() {
		CollectionDocument doc = HdfsMapFileDocumentResolver.smartDocumentResolver().loadCollectionDocument("<urn:uuid:da8b10cf-af5d-43c8-8cf1-ae2f1fba994e>");
		String expectedUrl = "http://en.wikipedia.org/wiki/Canis_lupus_familiaris";
		String expectedCanonicalUrl = "http://en.wikipedia.org/wiki/Dog";
		
		Assert.assertEquals(expectedUrl, doc.getUrl().toString());
		Assert.assertEquals(expectedCanonicalUrl, doc.getCanonicalUrl().toString());
	}
	
	@Test
	@SneakyThrows
	public void approveTransformationOfNonExistinClueWebDocumentClueWebDocumentFromHDFS() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Record-ID", "my-id-2");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		CollectionDocument doc = new HdfsMapFileDocumentResolver("cw09", "clueweb09").loadCollectionDocument("clueweb09-NON-EXISTING-en0000-00-00009");
		
		Assert.assertNull(doc);
	}
	
	@Test
	@SneakyThrows
	public void approveTransformationOfExistingCommonCrawlRecordWithMainContentExtraction() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Record-ID", "my-id-2");
		headers.put("WARC-Target-URI", "http://example.com");
		headers.put("WARC-Date", "01.01.1970");
		String documentContent = CollectionDocumentUtil.loadRawDocumentFromChatnoir(new URL("https://chatnoir.eu/cache?uuid=92105ce9-2938-5b89-a191-7a82fe5d8816&index=cc1704&raw"), CollectionDocumentUtil.RETRY_FINAL);
		WarcRecord record = record(headers, documentContent, "REsponse");
		
		CollectionDocument actual = transformToCollectionDocument(record, DocumentToTextTransformation.MAIN_CONTENT_EXTRACTION);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void approveTransformationOfClueWebRecordWithInvalidCase() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Trec-ID", "my-id-1");
		headers.put("WARC-target-URI", "http://example.com");
		headers.put("WARC-date", "01.01.1970");
		WarcRecord record = record(headers, "my-main-content", "response");
		
		CollectionDocument actual = transformToCollectionDocument(record);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void approveTransformationOfClueWebRecordWithResponseTypeRequest() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Trec-ID", "my-id-1");
		headers.put("WARC-target-URI", "http://example.com");
		headers.put("WARC-date", "01.01.1970");
		WarcRecord record = record(headers, "my-main-content", "request");
		
		CollectionDocument actual = transformToCollectionDocument(record);

		Assert.assertNull(actual);
	}
	
	@Test()
	public void approveTransformationOfClueWebRecordWithInvalidCaseAndInvalidURL() {
		Map<String, String> headers = new HashMap<>();
		headers.put("WARC-Trec-ID", "my-id-1");
		headers.put("WARC-target-URI", "example.com");
		headers.put("WARC-date", "01.01.1970");
		WarcRecord record = record(headers, "my-main-content", "resPonse");
		
		CollectionDocument actual = transformToCollectionDocument(record);
		
		Approvals.verifyAsJson(actual);
	}
	
	private CollectionDocument transformToCollectionDocument(WarcRecord record) {
		return transformToCollectionDocument(record, null);
	}
	
	private CollectionDocument transformToCollectionDocument(WarcRecord record, DocumentToTextTransformation transformation) {
		return CreateDocumentRepresentations.transformToCollectionDocument(record, transformation);
	}
	
	private static WarcRecord record(Map<String, String> headers, String body, String recordType) {
		WarcHeader header = Mockito.mock(WarcHeader.class);
		Mockito.when(header.getHeaderMetadata()).thenReturn(new TreeMap<>(headers));
		
		WarcRecord ret = Mockito.mock(WarcRecord.class);
		Mockito.when(ret.getContent()).thenReturn(body);
		Mockito.when(ret.getHeader()).thenReturn(header);
		Mockito.when(ret.getRecordType()).thenReturn(recordType);
		
		return ret;
	}
}
