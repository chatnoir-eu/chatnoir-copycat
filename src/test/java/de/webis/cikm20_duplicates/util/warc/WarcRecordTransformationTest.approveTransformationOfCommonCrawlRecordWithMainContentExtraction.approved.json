{
  "id": "my-id-2",
  "content": "",
  "fullyCanonicalizedContent": "",
  "crawlingTimestamp": "01.01.1970",
  "url": "http://example.com",
  "canonicalUrl": "https://example.com/test-123/"
}